### ISSUE TITLE
[Paper]: GLM-5: From Vibe Coding to Agentic Engineering

### ISSUE THREAD

**Issue by zyzkyrie** [kyriekevin/Latent-Atelier#40]
### Link / Source

https://arxiv.org/abs/2602.15763

**Comment by kyriekevin** [comment]
```json
{
  "summary": "GLM-5 is a 744B-parameter (40B active) MoE foundation model aimed at moving from prompt-driven “vibe coding” to long-horizon “agentic engineering,” emphasizing real software engineering and tool-using agents. The paper’s key system contributions are (i) adopting DeepSeek Sparse Attention (DSA) via continued pre-training to cut long-context attention cost while preserving fidelity up to 128K–200K contexts, (ii) building a fully asynchronous, decoupled RL post-training stack (slime-based) plus stability mechanisms (e.g., deterministic top-k for DSA indexer, token-in-token-out gateway, double-sided importance sampling/clipping) to scale long-horizon agent rollouts, and (iii) environment scaling recipes for verifiable SWE/terminal/search/slide-generation tasks with extensive benchmark coverage. Empirically, GLM-5 reports SOTA among open-weight models on multiple agentic/reasoning/coding benchmarks (e.g., SWE-bench Verified, BrowseComp w/ context management), plus internal “CC-Bench-V2” meant to approximate end-to-end engineering workflows. The work also documents substantial full-stack deployment/quantization/kernel optimizations for multiple Chinese GPU/NPU ecosystems, positioning GLM-5 as both a model release and an engineering playbook for efficient agentic RL at scale.",
  "strategic_review": {
    "A1_Significance": {
      "score": 4.4,
      "reason": "Targets two high-leverage bottlenecks simultaneously: (1) long-context cost/latency (DSA to reduce effective attention compute) and (2) post-training for long-horizon agents (async RL to eliminate rollout-induced GPU idle and improve autonomy). The emphasis on end-to-end software engineering (beyond static SWE-bench) is strategically important as agentic coding becomes a primary real-world workload."
    },
    "A2_Novelty": {
      "score": 3.8,
      "reason": "Most architectural novelty is borrowed (DSA, MLA/MTP lineage), but the paper provides a credible frontier-scale “system recipe” that combines continued pre-training for sparse attention, multi-stage RL (reasoning→agentic→general), and asynchronous RL infrastructure/stability fixes (TITO, deterministic top-k, double-sided IS/clipping, DP-aware routing). This is more than a tweak, but not a new primitive on the level of a broadly novel attention operator introduced here."
    },
    "A3_Alignment": {
      "score": 4.6,
      "reason": "Directly aligned with current critical-path needs for production agents: lowering long-context inference/training cost, improving long-horizon robustness, and increasing RL throughput via decoupled rollout/training engines. The paper’s concrete mechanisms (tail-latency controls, PD disaggregation, KV reuse routing) map cleanly onto near-term agent latency and reliability roadmaps."
    }
  },
  "tactical_review": {
    "B1_Rigor": {
      "score": 3.7,
      "reason": "Strong breadth of evaluations (reasoning, coding, agentic, long-horizon) and several targeted ablations (efficient attention variants; DSA warmup vs joint training; SWA pattern search). However, some key claims rely on internal/private components (private prompt sets for speculative decoding accept length, internal CC-Bench-V2, proprietary judges like o3-mini in BrowseComp standardization), and comparisons can be sensitive to agent harness details, timeouts, and judge choice—partially mitigated by documentation but still a reproducibility risk."
    },
    "B2_Transferability": {
      "score": 3.3,
      "reason": "Many ideas are portable across large-model training (async rollout/training decoupling, TITO vs re-tokenization mismatch, tail-latency engineering, cross-stage distillation), and the efficient-attention ablation insights generalize. Still, several gains are coupled to specific stack choices (DSA indexer behavior/top-k determinism, slime framework, MoE routing + KV-cache locality tricks, and extensive hardware-specific kernel fusion/quantization), making outcomes more conditional than universal across architectures/hardware."
    },
    "B3_Recipe_Detail": {
      "score": 4.4,
      "reason": "Unusually detailed engineering disclosure for a frontier-scale report: concrete token budgets and context schedule (32K/128K/200K), DSA adaptation steps (warmup steps, sequence length, LR), RL hyperparameters (e.g., \u03b2, \u03b5 ranges, group/batch sizes), and operational lessons (deterministic torch.topk; freezing indexer in RL; PD disaggregation; routing affinity). While not every dataset/judge is fully open, the text plus the linked GitHub substantially reduces “magic-number” guesswork."
    }
  }
}
```

**Comment by kyriekevin** [comment]
## Reading Map

### The Bullseye (Start Here)
* **Figure 5 (Overall training pipeline):** The “mechanism diagram” for the whole paper—pre-train → mid-train (context extension) → post-train (Reasoning RL → Agentic RL → General RL) + cross-stage distillation. Use this to anchor every later detail to *which stage caused which capability gain*.
* **Table 3 (MLA vs DSA long-context benchmarks):** Fastest way to validate the core architectural claim: DSA keeps long-context fidelity roughly on-par with dense attention while enabling cheaper compute. This is the key “accuracy didn’t fall off a cliff” evidence for sparse attention.
* **Table 6 (DSA warmup vs full joint training on RULER):** Tells you whether DSA is *practically adoptable* via continued training (indexer warmup + joint training) rather than a from-scratch rebuild. Also reveals where DSA still struggles (largest contexts) and how much training closes the gap.
* **Figure 8 (BrowseComp: keep-recent-\(k\) + Discard-all):** The cleanest, concrete “agent product” mechanism: context management is treated as an inference-time control knob that buys more steps and higher accuracy under fixed compute.
* **Figure 7 (Interleaved vs preserved thinking):** The behavioral mechanism for long-horizon coding agents—how they structure reasoning/tool calls across turns and how they attempt to prevent drift/inconsistency over long sessions.

### The Shortcuts (Ignore)
* Skip **Section 1 “Results” narrative around Figures 1–4** (mostly positioning/leaderboard context). Keep only the benchmark *names* you care about; return later if you need a headline number.
* Skim **Section 2.2 (Pre-training Data)** (standard corpus/filtering story). Only revisit if your question is “where did the coding/agent data come from?” or “did they avoid synthetic contamination?”
* Skim **Section 2.4 (Training Infrastructure)** unless you’re doing systems work. It’s valuable engineering, but not essential to understand *why* GLM-5 behaves more agentically.
* Skim **Section 3.4 (General RL)** (reward-model taxonomy is fairly standard). Focus instead on what’s unique/fragile: the async RL setup and the stability fixes.
* Skip **Section 5 (Chinese chip adaptation)** unless deployment on those accelerators is your goal. It’s a parallel contribution, largely orthogonal to the agentic-RL/DSA mechanism.
* Skip **Section 8 (Easter Eggs)** and **Section 9 (Contribution list)** for technical understanding.

### Delta Check
* [ ] **DSA causality:** Do **Table 3** and **Table 6** together actually justify “DSA is lossless by construction” in practice, or do they show residual degradation at the longest contexts that’s merely *reduced* by more training?
* [ ] **Async RL stability is real (not hand-wavy):** In **Section 3.2 (DSA RL insights)** and **Section 4.1.2**, do they provide evidence that determinism (**torch.topk**) + freezing the indexer + TITO + token-level clipping materially prevents collapse (e.g., entropy drop), or is it mostly anecdotal?
* [ ] **Benchmark fairness / evaluation knobs:** In **Section 6** (especially BrowseComp and tool-use tasks), how many results depend on *non-model* choices (judge model, official prompts, prompt adjustments, context management strategy, timeouts)? Would GLM-5 still look SOTA under a different judge or without those “evaluation hygiene” tweaks?
